### DISCUSSION TITLE
Cannot get computer by id on fabric

### DISCUSSION THREAD

**Discussion by carson-coder** [cc-tweaked/CC-Tweaked#1908] [General]
I am trying to get a computer by a id in my own mod with this code
```java
ServerContext server_ctx = ServerContext.get(Bellisimo.SERVER);
server_ctx.registry().get(id);
```
but vscode and gradle say that I cannot run get with a int even though https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.20.x/projects/common/src/main/java/dan200/computercraft/shared/computer/core/ServerComputerRegistry.java#L13 has that function. When decompiling the class in vscode the get(int) doesn't exist.

Full Error:
```
> Task :compileJava
/home/carson/bellisimo/src/main/java/com/rainbow/Register.java:61: error: incompatible types: int cannot be converted to UUID
                server_ctx.registry().get(id);
                                          ^
/home/carson/bellisimo/src/main/java/com/rainbow/Register.java:62: error: incompatible types: int cannot be converted to UUID
                Bellisimo.LOGGER.info(String.valueOf(server_ctx.registry().get(id).instanceID));
                                                                               ^
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
2 errors
```

**Comment by carson-coder** [comment]
I am on fabric 1.20.6 and computer craft 1.111.0

**Comment by zyxkad** [comment]
You should use
```java
ServerContext.get(<MinecraftServer instance>).registry().getComputers().stream().filter((computer) -> computer.getID() == numId).getFirst().orElse(null);
```

**Comment by SquidDev** [comment]
I think think this is a bit of an XY problem here. Mods should almost never be using `ServerComputer`, and should never look up a computer by its id — this is not guaranteed to be unique. What is the original problem you're trying to solve here?

**Reply by carson-coder** [reply]
I am trying to run a function on all instances of a peripheral that are on a certain computer. I need to get the computer by the id because this gets called by a command that has the computer id as a argument

**Reply by SquidDev** [reply]
> I need to get the computer by the id because this gets called by a command that has the computer id as a argument

Hrmr, this does feel like a bit of an anti-pattern. Peripherals tend to avoid indexing computers by id.

However, this can be pretty easily implemented by storing a map of computer id -> peripheral instance. I haven't tested this code, but something like the following should work:


```java
class MyPeripheral implements IPeripheral {
  private static final Map<Integer, Map<IComputerAccess, MyPeripheral>> attachedComputers = new HashMap<>();
  private static final Object lock = new Object();

  @Override
  public void attach(IComputerAccess computer) {
    synchronized(lock) {
      attachedComputers.computeIfAbsent(computer.getID(), HashMap::new).put(computer, this);
    }
  }

  @Override
  public void detach(IComputerAccess computer) {
    synchronized(lock) {
      var peripherals = attachedComputers.get(computer.getID());
      if (peripherals != null && peripherals.remove(computer) && peripherals.isEmpty()) {
        attachedComputers.remove(computer.getID())
      }
    }
  }

  public static forEachPeripheral(int id, BiConsumer<? super IComputerAccess, ? super MyPeripheral> action) {
    synchronized(lock) {
      // It's important we hold the lock while iterating over this collection,
      // hence writing this as a callback rather than returning an iterable.
      attachedComputers.getOrDefault(computer.getID()).forEach(action);
    }
  }
}
```

Also would recommend following the [Using section of the README](https://github.com/cc-tweaked/CC-Tweaked/tree/mc-1.20.y?tab=readme-ov-file#using) to set up your CC:T dependency. This should help you guard against using internal classes (like `ServerComputer`), which means your mod is less liable to break across CC versions.
